packages/vbridge/src/hooks/withdraw/useWithdraw.tsx

```diff
@@ -55,56 +64,125 @@ export const useWithdraw = (params: UseWithdrawProps) => {
+
+  const cancelWithdraw = useCallback(async () => {
+    if (canCancel) {
+      if (stage !== TransactionState.Ideal) {
+        await withdrawApi?.cancelWithdraw();
+        setStage(TransactionState.Ideal);
+      } else {
+        setStage(TransactionState.Ideal);
+      }
+    }
+  }, [canCancel, withdrawApi, stage, setStage]);
+
+  const setRelayer = useCallback(
+    (nextRelayer: WebbRelayer | null) => {
+      activeApi?.relayerManager.setActiveRelayer(nextRelayer, activeChain?.id!);
+    },
+    [activeApi?.relayerManager, activeChain]
+  );
+
   useEffect(() => {
-    const sub = activeApi?.relayerManager.listUpdated.subscribe(() => {
-      if (params.note) {
-        activeApi?.relayerManager.getRelayersByNote(params.note).then((r: WebbRelayer[]) => {
+    const subscriptions = params.notes?.map((note) =>
+      activeApi?.relayerManager.listUpdated.subscribe(() => {
+        if (!note) {
+          return;
+        }
+
+        activeApi?.relayerManager.getRelayersByNote(note).then((r: WebbRelayer[]) => {
           setRelayersState((p) => ({
             ...p,
             loading: false,
             relayers: r,
           }));
         });
-      }
-    });
-    return () => sub?.unsubscribe();
-  }, [activeApi, params.note, withdrawApi]);
-  const { registerInteractiveFeedback } = useWebContext();
+      })
+    );
+
+    return () => subscriptions?.forEach((sub) => sub?.unsubscribe());
+  }, [activeApi, params.notes, withdrawApi]);
+
   // hook events
   useEffect(() => {
-    if (params.note) {
-      activeApi?.relayerManager.getRelayersByNote(params.note).then((r: WebbRelayer[]) => {
+    params.notes?.map((note) =>
+      activeApi?.relayerManager.getRelayersByNote(note).then((r: WebbRelayer[]) => {
         setRelayersState((p) => ({
           ...p,
           loading: false,
           relayers: r,
         }));
-      });
-    }
+      })
+    );
 
     const sub = activeApi?.relayerManager.activeRelayerWatcher.subscribe((next: OptionalActiveRelayer) => {
       setRelayersState((p) => ({
         ...p,
         activeRelayer: next,
       }));
     });
-    const unsubscribe: Record<string, (() => void) | void> = {};
+
     if (!withdrawApi) {
       return;
     }
+
+    const unsubscribe: Record<string, (() => void) | void> = {};
     unsubscribe['stateChange'] = withdrawApi.on('stateChange', (stage: TransactionState) => {
       setStage(stage);
     });
 
     unsubscribe['validationError'] = withdrawApi.on(
       'validationError',
       (validationError: { note: string; recipient: string }) => {
-        setError((p) => ({
-          ...p,
-          validationError,
-        }));
+        setError((p) => {
+          const noteErrors = p.validationError.notes.slice();
+          noteErrors.push(validationError.note);
+
+          return {
+            ...p,
+            validationError: {
+              notes: noteErrors,
+              recipient: validationError.recipient,
+            },
+          };
```

Comment:
I'm not sure whether it's good or not when I handle note errors like this. Don't know how to test by the way 😢

Reply:
@AtelyPham  it's okay we can spec that next